### PR TITLE
add CombinatorialGapKFold

### PR DESCRIPTION
**Pull request by aldder**
From "_Advances in Financial Machine Learning_" book by Marcos López de Prado
the implemented version of Combinatorial Cross Validation with Purging and Embargoing

image

explaining video: https://www.youtube.com/watch?v=hDQssGntmFA

### CODE REVIEW

**Comment by WenjieZ** [comment]
Hi @aldder , please try to add some test cases in the `test_split.py` file.

File: tscv/_split.py [hunk 1]
```diff
@@ -371,6 +372,128 @@ def get_n_splits(self, X=None, y=None, groups=None):
...
+            raise ValueError(
+                "Combinatorial k-fold cross-validation requires at least one"
+                " test split by setting k=1 or more,"
+                " got k={0}.".format(k))
+
+        super().__init__(gap_before, gap_after)
+        self.n_groups = N
+        self.test_splits = k
+
+    def split(self, X, y=None, groups=None):
```
Comment (WenjieZ) [inline tscv/_split.py:447]: The canonical way of doing this is to redefine `_iter_test_indices(self, X, y=None, groups=None)` from the base class and generate only the test indices. The base class will take care of the rest. Please refer to the other derived classes and make modification accordingly.

File: tscv/_split.py [hunk 2]
```diff
@@ -371,6 +372,128 @@ def get_n_splits(self, X=None, y=None, groups=None):
...
+
+        groups : object
+            Always ignored, exists for compatibility.
+
+        Returns
+        -------
+        n_splits : int
+            Returns the number of splitting iterations in the cross-validator.
+        """
+        return len(list(combinations(range(self.n_groups), self.test_splits)))
```
Comment (WenjieZ) [inline tscv/_split.py:494]: Use the combination number to generate the result directly rather than instantiating all combinations.

**Review by WenjieZ — CHANGES_REQUESTED** [review]
Please see the comments and make changes accordingly.

File: tscv/_split.py (hunk 1 above)
Comment (WenjieZ) [reply tscv/_split.py:448]: The check of the training set size can be delayed and implemented in `GapCrossValidator._iter_train_indices()` and its 3 siblings. You can also implement it in `GapCrossValidator.split()` if you find it a hustle to implement it four times.

File: tscv/_split.py [hunk 3]
```diff
@@ -10,8 +10,9 @@
 import numbers
 from math import modf
 from abc import ABCMeta, abstractmethod
-from itertools import chain
+from itertools import chain, combinations
 from inspect import signature
+from scipy.special import comb
```
Comment (WenjieZ) [inline tscv/_split.py:15]: Please add `scipy` in https://github.com/WenjieZ/TSCV/blob/2abbc3d34151555ea12c9abb13d7a76ead5a20a8/setup.py#L55

File: tscv/_split.py (hunk 1 above)
Comment (WenjieZ) [reply tscv/_split.py:448]: On second thought, I don't find it necessary to check the non-emptiness of the training set. Some models/algorithms can output a default estimator/strategy given an empty training set (e.g., equal weight portfolio). If your model/algorithm requires the non-empty training set, it's probably a good idea to check it in the model/algorithm rather than in the cross-validator. That's why I didn't implement this check when I released the package. 

That said, I probably forgot to check the non-emptiness of the test set.

Comment (aldder) [reply tscv/_split.py:448]: Ok then, I will replace the implementation of the split method with _iter_test_indices

About the non-emptiness of the test set I think you prefer to implement it in the base class with a specific PR, so I won't touch it, ok?

Comment (WenjieZ) [reply tscv/_split.py]: Your judgement is correct.
